recipes/default.rb

```diff
@@ -63,28 +68,74 @@
   action [:start, :enable]
 end
 
-user "jenkins" do
+user agent_username  do
   shell "/bin/bash"
   manage_home true
 end
-sudo "jenkins" do
-  user "jenkins"
+sudo agent_username do
+  user agent_username
   nopasswd true
 end
 
 # Add agent user to the docker group to allow them to build and run docker
 # containers.
 group 'docker' do
   append true
-  members 'jenkins'
+  members agent_username
   action :manage # Group should be created by docker package.
 end
 
-directory "/home/jenkins/jenkins-agent"
-agent_jar_url = node["osrf_jenkins_agent"]["agent_jar_url"]
-if agent_jar_url.nil? || agent_jar_url.empty?
-  agent_jar_url = "#{node["osrf_jenkins_agent"]["jenkins_url"]}/jnlpJars/agent.jar"
+
+# TODO: how to read attributes from chef-osrf plugins into this cookbook
+# swarm_client_version = node['jenkins-plugins']['swarm']
+swarm_client_version = "3.24"
+swarm_client_url = "https://repo.jenkins-ci.org/releases/org/jenkins-ci/plugins/swarm-client/#{swarm_client_version}/swarm-client-#{swarm_client_version}.jar"
+swarm_client_jarfile_path = "/home/#{agent_username}/swarm-client-#{swarm_client_version}.jar"
+
+# Download swarm client program from url and install it to the jenkins-agent user's home directory.
+remote_file swarm_client_jarfile_path do
+  source swarm_client_url
+  owner agent_username
+  group agent_username
+  mode '0444'
+end
+
+jenkins_username = node['osrf_buildfarm']['agent']['username']
+agent_jenkins_user = search('osrf_buildfarm_jenkins_users', "username:#{jenkins_username}").first
+template '/etc/default/jenkins-agent' do
+  source 'jenkins-agent.env.erb'
+  variables Hash[
+    java_args: node['osrf_buildfarm']['agent']['java_args'],
+    jarfile: swarm_client_jarfile_path,
+    jenkins_url: node['osrf_buildfarm']['jenkins_url'],
+    username: jenkins_username,
+    password: agent_jenkins_user['password'],
+    name: node['osrf_buildfarm']['agent']['nodename'],
+    description: node['osrf_buildfarm']['agent']['description'],
+    executors: node['osrf_buildfarm']['agent']['executors'],
+    user_home: agent_homedir,
+    labels: node['osrf_buildfarm']['agent']['labels'],
+  ]
+  notifies :restart, 'service[jenkins-agent]'
 end
-remote_file "/home/jenkins/jenkins-agent/agent.jar" do
-  source agent_jar_url
+
+template '/etc/systemd/system/jenkins-agent.service' do
+  source 'jenkins-agent.service.erb'
+  variables Hash[
+    service_name: 'jenkins-agent',
+    username: agent_username,
+  ]
+  notifies :run, 'execute[systemctl-daemon-reload]', :immediately
+  notifies :restart, 'service[jenkins-agent]'
+end
+
+execute 'systemctl-daemon-reload' do
+  command 'systemctl daemon-reload'
+  action :nothing
+end
+
+service 'jenkins-agent' do
+  action [:start, :enable]
+  # can not connect to server while testing
+  not_if { node.chef_environment == "test" }
```

Comment:
I agree with you here. What you see wrong in the code about avoiding the `jenkins-agent` service in the case of running test environment?

Reply:
The not_if here won't try to start the jenkins-agent service in test environments. But I think it should start it in all cases eve if you can't connect to the server.
```suggestion
```

Reply:
We can try, I think that it is going to make CI to fail since the agent connection will return an error code from provisioning making the provision to stop there. Let's see if that is correct.

Reply:
>  Let's see if that is correct.

Failing now after applying the change. 
```

  * service[jenkins-agent] action restart  * service[jenkins-agent] action restart
    
    ================================================================================
    Error executing action `restart` on resource 'service[jenkins-agent]'
    ================================================================================
    
    Mixlib::ShellOut::ShellCommandFailed
    ------------------------------------
    Expected process to exit with [0], but received '1'
    ---- Begin output of /usr/bin/systemctl --system restart jenkins-agent ----
    STDOUT: 
    STDERR: Job for jenkins-agent.service failed because the control process exited with error code.
    See "systemctl status jenkins-agent.service" and "journalctl -xe" for details.
    ---- End output of /usr/bin/systemctl --system restart jenkins-agent ----
    Ran /usr/bin/systemctl --system restart jenkins-agent returned 1
```

Reply:
I'm going to revert the change to make CI happy. Ticketed in #7 in the case we find time to debug and solve it.